apps/kg/kvserver.py

```diff
@@ -0,0 +1,149 @@
+                          help='the format of the dataset.')
+
+        self.add_argument('--hidden_dim', type=int, default=256,
+                          help='hidden dim used by relation and entity')
+        self.add_argument('--lr', type=float, default=0.0001,
+                          help='learning rate')
+        self.add_argument('-g', '--gamma', type=float, default=12.0,
+                          help='margin value')
+
+        self.add_argument('--gpu', type=int, default=[-1], nargs='+',
+                          help='a list of active gpu ids, e.g. 0')
+        self.add_argument('--mix_cpu_gpu', action='store_true',
+                          help='mix CPU and GPU training')
+        self.add_argument('-de', '--double_ent', action='store_true',
+                          help='double entitiy dim for complex number')
+        self.add_argument('-dr', '--double_rel', action='store_true',
+                          help='double relation dim for complex number')
+        self.add_argument('--seed', type=int, default=0,
+                          help='set random seed for reproducibility')
+
+        self.add_argument('--rel_part', action='store_true',
+                          help='enable relation partitioning')
+        self.add_argument('--soft_rel_part', action='store_true',
+                          help='enable soft relation partition')
+        self.add_argument('--nomp_thread_per_process', type=int, default=-1,
+                          help='num of omp threads used per process in multi-process training')
+        self.add_argument('--async_update', action='store_true',
+                          help='allow async_update on node embedding')
+        self.add_argument('--strict_rel_part', action='store_true',
+                          help='Strict relation partition')
+
+        self.add_argument('--server_id', type=int, default=0,
+                          help='Unique ID of each server')
+        self.add_argument('--ip_config', type=str, default='ip_config.txt',
+                          help='IP configuration file of kvstore')
+        self.add_argument('--total_client', type=int, default=1,
+                          help='Total number of client worker nodes')
+
+
+def get_server_data(args, machine_id):
+   """Get data from data_path/dataset/part_machine_id
+
+      Return: glocal2local, 
+              entity_emb, 
+              entity_state, 
+              relation_emb, 
+              relation_emb_state
+   """
+   g2l, dataset = get_server_partition_dataset(
+    args.data_path, 
+    args.dataset, 
+    args.format, 
+    machine_id)
+
+   # Note that the dataset doesn't ccontain the triple
+   print('n_entities: ' + str(dataset.n_entities))
+   print('n_relations: ' + str(dataset.n_relations))
+
+   model = load_model(None, args, dataset.n_entities, dataset.n_relations)
+
+   return g2l, model.entity_emb.emb, model.entity_emb.state_sum, model.relation_emb.emb, model.relation_emb.state_sum
```

Comment:
In this case, some model like transR are not supported by distributed training. 
They require additional embeddings.

Reply:
Yes, it is.